apps/common/lib/lexical/ast/analysis/analyzer.ex

```diff
@@ -0,0 +1,424 @@
+defmodule Lexical.Ast.Analysis.Analyzer do
+  @moduledoc false
+
+  alias __MODULE__
+  alias Lexical.Document
+  alias Lexical.Document.Position
+  alias Lexical.Document.Range
+  alias Sourceror.Zipper
+
+  @scope_id :_scope_id
+
+  @block_keywords [:do, :else, :rescue, :catch, :after]
+  @clauses [:->]
+
+  defmodule Alias do
+    defstruct [:module, :as, :line]
+
+    def new(module, as, line) when is_list(module) and is_atom(as) and line > 0 do
+      %Alias{module: module, as: as, line: line}
+    end
+  end
+
+  defmodule Scope do
+    defstruct [:id, :range, module: [], parent_aliases: %{}, aliases: []]
+
+    def new(id, %Range{} = range, parent_aliases \\ %{}, module \\ []) do
+      %Scope{id: id, range: range, module: module, parent_aliases: parent_aliases}
+    end
+
+    def global do
+      %Scope{id: :global}
+    end
+
+    def alias_map(%Scope{} = scope) do
+      scope.aliases
+      |> Enum.sort_by(& &1.line)
```

Comment:
we should write a comment about this sort before mapping. it looks wrong, but I think that's just to ensure that later aliases in a given scope win.

Reply:
Will add a comment and this also made me realize that there was a bug as I was using this function incorrectly in my implementation of `Analysis.aliases_at`.